### PR TITLE
[Vertex AI] Align text and add close keyboard function in Sample Project

### PR DESCRIPTION
**Pull request by 1998code**
### Discussion
- The keyboard should be hidden automatically in the ```Text``` and ```Multi-modal``` sections to prevent blocking content.
- The text and input field are not aligned in ```Text``` section.
- The input field sticks directly on the keyboard without any padding in ```Chat``` and ``Function Calling``` sections.

### Testing
Just alignment and padding updated. Everything is working fine.

### API Changes
N/A

### CODE REVIEW

**Comment by 1998code** [comment]
Signed

**Comment by 1998code** [comment]
Rerun

**Comment by 1998code** [comment]
🙏

**Comment by paulb777** [comment]
Thanks for the PR @1998code! We should review later this week.

File: FirebaseVertexAI/Sample/GenerativeAITextSample/Screens/SummarizeScreen.swift [hunk 1]
```diff
@@ -61,6 +64,15 @@ struct SummarizeScreen: View {
...
 
   private func onSummarizeTapped() {
+    // MARK: Close Keyboard
+
+    UIApplication.shared.sendAction(
+      #selector(UIResponder.resignFirstResponder),
+      to: nil,
+      from: nil,
+      for: nil
+    )
```
Comment (andrewheard) [inline FirebaseVertexAI/Sample/GenerativeAITextSample/Screens/SummarizeScreen.swift:74]: With the above [`focused(_:equals:)`](https://developer.apple.com/documentation/swiftui/view/focused(_:equals:)), I _think_ you could set `focusedField = nil` here to close the keyboard instead.

File: FirebaseVertexAI/Sample/GenerativeAITextSample/Screens/SummarizeScreen.swift [hunk 2]
```diff
@@ -28,19 +28,22 @@ struct SummarizeScreen: View {
...
-      Text("Enter some text, then tap on _Go_ to summarize it.")
-      HStack(alignment: .top) {
-        TextField("Enter text summarize", text: $userInput, axis: .vertical)
-          .textFieldStyle(.roundedBorder)
-          .onSubmit {
+      VStack(alignment: .leading) {
+        Text("Enter some text, then tap on _Go_ to summarize it.")
+          .padding(.horizontal, 6)
+        HStack(alignment: .top) {
+          TextField("Enter text summarize", text: $userInput, axis: .vertical)
```
Comment (andrewheard) [inline FirebaseVertexAI/Sample/GenerativeAITextSample/Screens/SummarizeScreen.swift:35]: I think it might be better to manage the [focus](https://developer.apple.com/documentation/swiftui/focusstate) within the view instead of explicitly closing the keyboard. This view actually already has a property `focusedField` for this but it was never hooked up. Adding `.focused($focusedField, equals: .message)` to the `TextField` should automatically modify the state when the user taps in the field.

**Review by andrewheard — CHANGES_REQUESTED** [review]
Thank you very much, @1998code! I haven't had a chance to review the PR in detail yet but added a couple comments about the keyboard closing.

By the way, I tried running the Chat and Function Calling samples and they look much better with the additional padding!

File: FirebaseVertexAI/Sample/GenerativeAITextSample/Screens/SummarizeScreen.swift (hunk 1 above)
Comment (1998code) [reply FirebaseVertexAI/Sample/GenerativeAITextSample/Screens/SummarizeScreen.swift:74]: Yes, I did try ```focusedField = nil``` before adding this line of code.
However, it is not working on my side 🤔

image

File: FirebaseVertexAI/Sample/GenerativeAITextSample/Screens/SummarizeScreen.swift (hunk 2 above)
Comment (1998code) [reply FirebaseVertexAI/Sample/GenerativeAITextSample/Screens/SummarizeScreen.swift:35]: Yes, you are right. Adding now.

File: FirebaseVertexAI/Sample/ChatSample/Screens/ConversationScreen.swift [hunk 3]
```diff
@@ -68,9 +68,11 @@ struct ConversationScreen: View {
       InputField("Message...", text: $userPrompt) {
         Image(systemName: viewModel.busy ? "stop.circle.fill" : "arrow.up.circle.fill")
           .font(.title)
+          .padding(.bottom, 3)
```
Comment (peterfriese) [inline FirebaseVertexAI/Sample/ChatSample/Screens/ConversationScreen.swift:71]: Instead of padding here, we should fix the padding on the InputField.


Here is how the `body` function in `InputField` should be updated:

```
  public var body: some View {
    VStack(alignment: .leading) {
      HStack(alignment: .bottom) {
        VStack(alignment: .leading) {
          TextField(
            title ?? "",
            text: $text,
            axis: .vertical
          )
          .padding(.vertical, 4)
          .onSubmit(submit)
        }
        .padding(.horizontal, 8)
        .padding(.vertical, 4)
        .overlay {
          RoundedRectangle(
            cornerRadius: 8,
            style: .continuous
          )
          .stroke(Color(UIColor.systemFill), lineWidth: 1)
        }

        Button(action: submit, label: label)
          .padding(.bottom, 4) // .bottom instead of .top
      }
    }
    .padding(8) // 8 for both horizontal and vertical instead of default for just horizontal
  }
}
```


Note that only the lines with comments need to be changed, everything else can stay the same.

File: FirebaseVertexAI/Sample/ChatSample/Screens/ConversationScreen.swift [hunk 4]
```diff
@@ -68,9 +68,11 @@ struct ConversationScreen: View {
...
       }
       .focused($focusedField, equals: .message)
       .onSubmit { sendOrStop() }
+      .padding(.bottom, 5)
```
Comment (peterfriese) [inline FirebaseVertexAI/Sample/ChatSample/Screens/ConversationScreen.swift:75]: Instead of padding here, we should fix the padding on the InputField.

File: FirebaseVertexAI/Sample/FunctionCallingSample/Screens/FunctionCallingScreen.swift [hunk 5]
```diff
@@ -69,9 +69,11 @@ struct FunctionCallingScreen: View {
       InputField("Message...", text: $userPrompt) {
         Image(systemName: viewModel.busy ? "stop.circle.fill" : "arrow.up.circle.fill")
           .font(.title)
+          .padding(.bottom, 3)
```
Comment (peterfriese) [inline FirebaseVertexAI/Sample/FunctionCallingSample/Screens/FunctionCallingScreen.swift:72]: Instead of padding here, we should fix the padding on the InputField.

File: FirebaseVertexAI/Sample/FunctionCallingSample/Screens/FunctionCallingScreen.swift [hunk 6]
```diff
@@ -69,9 +69,11 @@ struct FunctionCallingScreen: View {
...
       }
       .focused($focusedField, equals: .message)
       .onSubmit { sendOrStop() }
+      .padding(.bottom, 5)
```
Comment (peterfriese) [inline FirebaseVertexAI/Sample/FunctionCallingSample/Screens/FunctionCallingScreen.swift:76]: Instead of padding here, we should fix the padding on the InputField.

File: FirebaseVertexAI/Sample/GenerativeAITextSample/Screens/SummarizeScreen.swift [hunk 7]
```diff
@@ -61,6 +65,10 @@ struct SummarizeScreen: View {
   }
 
   private func onSummarizeTapped() {
+    // MARK: Close Keyboard
```
Comment (peterfriese) [inline FirebaseVertexAI/Sample/GenerativeAITextSample/Screens/SummarizeScreen.swift:68]: MARK comments should be used to separate parts of the code (see https://www.avanderlee.com/xcode/xcode-mark-line-comment/).

I think we don't need a comment here, as focus handling should be a well-known technique in the community.

File: FirebaseVertexAI/Sample/GenerativeAITextSample/Screens/SummarizeScreen.swift [hunk 8]
```diff
@@ -61,6 +65,10 @@ struct SummarizeScreen: View {
...
+
```
Comment (peterfriese) [inline FirebaseVertexAI/Sample/GenerativeAITextSample/Screens/SummarizeScreen.swift:69]: Nit: remove empty line

**Review by peterfriese — CHANGES_REQUESTED** [review]
Thanks for looking into this. I think the paddings should be implemented on the InputField instead.

Also, if you've got time, can you look into adding focus handling to the multi modal sample as well (`PhotoReasoningScreen`)?

File: FirebaseVertexAI/Sample/ChatSample/Screens/ConversationScreen.swift (hunk 3 above)
Comment (1998code) [reply FirebaseVertexAI/Sample/ChatSample/Screens/ConversationScreen.swift]: Updated

File: FirebaseVertexAI/Sample/ChatSample/Screens/ConversationScreen.swift (hunk 4 above)
Comment (1998code) [reply FirebaseVertexAI/Sample/ChatSample/Screens/ConversationScreen.swift]: Removed 🙂

File: FirebaseVertexAI/Sample/FunctionCallingSample/Screens/FunctionCallingScreen.swift (hunk 6 above)
Comment (1998code) [reply FirebaseVertexAI/Sample/FunctionCallingSample/Screens/FunctionCallingScreen.swift]: Fixed

File: FirebaseVertexAI/Sample/GenerativeAITextSample/Screens/SummarizeScreen.swift (hunk 7 above)
Comment (1998code) [reply FirebaseVertexAI/Sample/GenerativeAITextSample/Screens/SummarizeScreen.swift]: Removed 🙏

File: FirebaseVertexAI/Sample/GenerativeAITextSample/Screens/SummarizeScreen.swift (hunk 8 above)
Comment (1998code) [reply FirebaseVertexAI/Sample/GenerativeAITextSample/Screens/SummarizeScreen.swift]: Removed too. The empty line is added by the style script 🙈

**Comment by 1998code** [comment]
@peterfriese I added focus handling to PhotoReasoningScreen too.

**Comment by 1998code** [comment]
[ Function Calling Screen ]
Btw I noticed the message was placed on ```user``` side instead of ```system```.

| Before | After |
| - | - |
| IMG_F9AD8B383DF0-1 | 螢幕截圖 2024-05-10 下午8 31 31 |

**Comment by 1998code** [comment]
Thanks.

File: FirebaseVertexAI/Sample/GenerativeAIUIComponents/Sources/GenerativeAIUIComponents/InputField.swift [hunk 9]
```diff
@@ -38,32 +38,32 @@ public struct InputField<Label>: View where Label: View {
...
   public var body: some View {
-    VStack(alignment: .leading) {
-      HStack(alignment: .bottom) {
-        VStack(alignment: .leading) {
-          TextField(
-            title ?? "",
-            text: $text,
-            axis: .vertical
-          )
+      VStack(alignment: .leading) {
```
Comment (peterfriese) [inline FirebaseVertexAI/Sample/GenerativeAIUIComponents/Sources/GenerativeAIUIComponents/InputField.swift:41]: Nit: Please use an indent of 2 spaces.

Comment (1998code) [reply FirebaseVertexAI/Sample/GenerativeAIUIComponents/Sources/GenerativeAIUIComponents/InputField.swift]: Updated 🙏

File: FirebaseVertexAI/Sample/FunctionCallingSample/Screens/FunctionCallingScreen.swift [hunk 10]
```diff
@@ -65,6 +65,9 @@ struct FunctionCallingScreen: View {
             }
           }
         })
+        .onTapGesture {
+          focusedField = nil
+        }
```
Comment (andrewheard) [inline FirebaseVertexAI/Sample/FunctionCallingSample/Screens/FunctionCallingScreen.swift:70]: Thanks, I like how this closes the keyboard if you tap outside the text field!

**Review by andrewheard — CHANGES_REQUESTED** [review]
Thanks, @1998code! If it's not too much trouble, would you be able to rebase your PR on `main` instead of `vertex-ai`? We've merged Vertex AI into https://github.com/firebase/firebase-ios-sdk/tree/main/FirebaseVertexAI/Sample and are no longer using the `vertex-ai` branch. (If it's too much trouble we can merge here and copy it into `main` after.)

@peterfriese Do you have any other concerns? This LGTM.

File: FirebaseVertexAI/Sample/FunctionCallingSample/ViewModels/FunctionCallingViewModel.swift [hunk 11]
```diff
@@ -249,7 +249,7 @@ private extension FunctionResponse {
     }
     let messageText = "Function response returned by app:\n```\n\(json)\n```"
 
-    return ChatMessage(message: messageText, participant: .user)
+    return ChatMessage(message: messageText, participant: .system)
```
[the remark this answers is not in the document]
Comment (1998code) [reply FirebaseVertexAI/Sample/FunctionCallingSample/ViewModels/FunctionCallingViewModel.swift]: Okay. Reverted

File: FirebaseVertexAI/Sample/FunctionCallingSample/Screens/FunctionCallingScreen.swift (hunk 10 above)
Comment (1998code) [reply FirebaseVertexAI/Sample/FunctionCallingSample/Screens/FunctionCallingScreen.swift:70]: 🙏

**Comment by 1998code** [comment]
Rebased 🙏

**Comment by 1998code** [comment]
Screenshot 2024-05-14 at 10 25 18 AM

CLA checking seems failed.

**Comment by paulb777** [comment]
Looks like we don't have a clean rebase

**Comment by 1998code** [comment]
🤔 Should we simply copy the updated code and open a new PR?

**Comment by paulb777** [comment]
Maybe. There is a way to do it in git with `git rebase --onto` and another parameter, but either way is fine.

**Comment by 1998code** [comment]
Okay. Let me try again.

**Comment by 1998code** [comment]
Please see if everything is okay 🙏

File: FirebaseVertexAI/Sources/GenerativeAISwift.swift
Comment (andrewheard) [inline FirebaseVertexAI/Sources/GenerativeAISwift.swift:1]: Looks like this file was accidentally added back during the rebase. Sorry for the hassle!

**Review by andrewheard — CHANGES_REQUESTED** [review]
Thanks, @1998code! Other than the re-added `GenerativeAISwift.swift` file this LGTM.

File: FirebaseVertexAI/Sources/GenerativeAISwift.swift
Comment (1998code) [reply FirebaseVertexAI/Sources/GenerativeAISwift.swift:1]: Removed 🙏

**Review by peterfriese — APPROVED** [review]
LGTM, thanks for your contribution!

**Review by andrewheard — APPROVED** [review]
Thank you for your contribution (and patience with our reviews), @1998code!

**Comment by 1998code** [comment]
Thanks all. Happy GoogleIO day ♥️

**Comment by andrewheard** [comment]
@1998code If you're interested in making a contribution in another repository, we have _nearly_ the same sample app for the Google AI SDK: https://github.com/google-gemini/generative-ai-swift/tree/main/Examples/GenerativeAISample

No pressure though, I'd be happy to copy your improvements there.

**Comment by 1998code** [comment]
Updated 😊 @andrewheard 
Please review https://github.com/google-gemini/generative-ai-swift/pull/172
